recipes-openxt/openxt/openxt-measuredlaunch/ml-functions

```diff
@@ -0,0 +1,375 @@
+            ;;
+            *)
+                break
+            ;;
+            esac
+
+            mod_num=$(( ${mod_num} + 1 ))
+        done
+
+        [ -n ${pcr18} ] && {
+            echo "${pcr18}"
+            return 0
+        }
+    fi
+
+    return 1
+}
+
+# Function to calculate PCR 19
+# parameter 0: path to grub.cfg
+# return 0 on success
+# return 1 on failure
+pcr19() {
+    local root="${1:-/}"
+    local grub_conf=${2}
+    local pcr19=''
+
+    modules=$(parse_grub ${root}/$grub_conf)
+
+    if [[ $? -eq 0 ]]; then
+        local IFS=$'\n'
+        mod_num=0
+        for  modline in $modules; do
+            args=""
+            module=$(echo ${modline} | awk '{ print $1 }')
+            if [[ $(echo ${modline} | wc -w) -gt 2 ]]; then
+                cmdline=$(echo ${modline} | awk '{ print substr($0, index($0,$2)) }')
+            else
+                cmdline=""
+            fi
+
+            if $(echo ${module} | egrep -q -e '\.bin$|\.BIN$'); then
+                finger=$(head -c 8 ${root}/${module}|sha1sum|head -c 7)
+                if [ "$finger" = "c22c443" ]; then
+                    continue
+                fi
+            fi
+
+            case ${mod_num} in
+            0)
+                mod_num=$(( ${mod_num} + 1 ))
+                continue
+            ;;
+            1)
+                mod_num=$(( ${mod_num} + 1 ))
+                continue
+            ;;
+            *)
+                [[ -z ${pcr19} ]] && pcr19=0
+
+                [[ -n "${cmdline}" ]] && args="-c '${cmdline}'"
+                $(echo ${module}|egrep -q -e'\.gz$') && args="${args} -z"
+                pcr19=$(eval "module_hash -e ${pcr19} ${args} ${root}/${module}")
+            ;;
+            esac
+
+            mod_num=$(( ${mod_num} + 1 ))
+        done
+
+        [ -n ${pcr19} ] && {
+            echo "${pcr19}"
+            return 0
+        }
+    fi
+
+    return 1
+}
+
+calculate_pcrs() {
+    local root="${1:-/}"
+    local grub_conf=${2}
+	local p17=''
+	local p18=''
+	local p19=''
+
+	p17=$(pcr17 ${root} ${grub_conf} "true") || return 1
+	p18=$(pcr18 ${root} ${grub_conf}) || return 1
+	p19=$(pcr19 ${root} ${grub_conf}) || return 1
+
+	echo "${p17} ${p18} ${p19}"
+}
+
+# Function to generate the tboot verified launch policy
+# parameter 0: TPM owner password
+# return 0 on success
+# return 1 on failure to create policy
+# return 2 on failure to write the policy to the TPM
+generate_policy() {
+    local password=$(cat $1)
+    local policy=/tmp/tboot.vl
+    local tboot_idx='0x20000001'
+
+    [ -f $policy ] && rm -f $policy
+
+    tb_polgen --create --type nonfatal --ctrl 1 $policy 2>&1 1>/dev/null && \
+    tb_polgen --add --num 0 --pcr none --hash any $policy 2>&1 1>/dev/null && \
+    tb_polgen --add --num any --pcr 19 --hash any $policy 2>&1 1>/dev/null
+    if [ $? -ne 0 ]; then
+        echo "failed to generate tboot Verified Launch policy" >&2
+        return 1
```

Comment:
Its only a temp file with no sensitive information. The reason it removed at beginning is that tb_polgen --create fails if the file exist, IIRC.